pageserver/src/walredo.rs

```diff
@@ -880,14 +852,95 @@ impl PostgresRedoProcess {
     }
 }
 
-impl Drop for PostgresRedoProcess {
-    fn drop(&mut self) {
-        if !self.called_kill {
-            error!(tenant_id=%self.tenant_id, pid = %self.child.id(), "dropping PostgresRedoProcess that wasn't killed, likely causing defunct postgres process");
+/// Wrapper around st::process::Child that provides a
+struct NoLeakChild {
+    child: Option<Child>,
+}
+
+/// Wrapper type around `std::process::Child` which guarantees that the child
+/// will be killed and waited-for by this proces before being dropped.
+///
+impl NoLeakChild {
+    fn spawn(command: &mut Command) -> io::Result<Self> {
+        let child = command.spawn()?;
+        Ok(NoLeakChild { child: Some(child) })
+    }
+
+    fn get_mut_child(&mut self) -> &mut Child {
+        self.child.as_mut().expect("must not call from drop")
+    }
+
```

Comment:
How about implementing `Deref` and `DerefMut` for `Child` instead? I think that would eliminate the wrapper for `id` too.